File: exp-14/q1-kruskals/solution.py
import sys

def find(parent, x):
    if parent[x] != x:
        parent[x] = find(parent, parent[x])
    return parent[x]

def main():
    data = sys.stdin.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1

    edges = []
    for _ in range(m):
        u = int(data[idx]); idx += 1
        v = int(data[idx]); idx += 1
        w = int(data[idx]); idx += 1
        edges.append((w, u, v))

    edges.sort()

    parent = list(range(n + 1))

    print("Edges in MST:")
    total = 0
    count = 0
    for w, u, v in edges:
        if count == n - 1:
            break
        ru = find(parent, u)
        rv = find(parent, v)
        if ru != rv:
            parent[ru] = rv
            print(f"{u} - {v} ({w})")
            total += w
            count += 1

    print(f"Total weight of MST: {total}")

if __name__ == "__main__":
    main()

File: exp-14/q1-kruskals/solution.js
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function find(parent, x) {
    if (parent[x] !== x)
        parent[x] = find(parent, parent[x]);
    return parent[x];
}

const lines = [];
rl.on('line', (line) => lines.push(line.trim()));
rl.on('close', () => {
    const nums = lines.join(' ').split(/\s+/).map(Number);
    let idx = 0;
    const n = nums[idx++];
    const m = nums[idx++];

    const edges = [];
    for (let i = 0; i < m; i++) {
        const u = nums[idx++];
        const v = nums[idx++];
        const w = nums[idx++];
        edges.push({ u, v, w });
    }

    edges.sort((a, b) => a.w - b.w);

    const parent = Array.from({ length: n + 1 }, (_, i) => i);

    console.log('Edges in MST:');
    let total = 0, count = 0;
    for (const e of edges) {
        if (count === n - 1) break;
        const ru = find(parent, e.u);
        const rv = find(parent, e.v);
        if (ru !== rv) {
            parent[ru] = rv;
            console.log(`${e.u} - ${e.v} (${e.w})`);
            total += e.w;
            count++;
        }
    }

    console.log(`Total weight of MST: ${total}`);
});
